gcc/rust/resolve/rust-ast-resolve-pattern.cc

```diff
@@ -23,6 +23,71 @@
 namespace Rust {
 namespace Resolver {
 
+void
+PatternDeclaration::go (AST::Pattern *pattern, Rib::ItemType type)
+{
+  std::vector<PatternBinding> bindings
+    = {std::make_pair (PatternBoundCtx::Product, std::set<Identifier> ())};
+  PatternDeclaration resolver (&bindings, type);
+  pattern->accept_vis (resolver);
+};
+
+void
+PatternDeclaration::go (AST::Pattern *pattern,
+			std::vector<PatternBinding> *bindings,
+			Rib::ItemType type)
+{
+  PatternDeclaration resolver (bindings, type);
+  pattern->accept_vis (resolver);
+}
+
+void
+PatternDeclaration::visit (AST::IdentifierPattern &pattern)
+{
+  rust_assert (!bindings->empty ());
+  if (bindings->back ().first == PatternBoundCtx::Product
+      && bindings->back ().second.find (pattern.get_ident ())
```

Comment:
This if statement is not readable. If your doing something like this, you need to check for:

```
bool have_bindings = bindings.size() > 0;
bool have_product_bindings = have_bindings && bindings.back().first == PatternBoundCtx::Product;
...

if (have_product_bindings && .. )
```

I am not sure about this pattern why is this a vector? because you are only checking binds.back, I don't see any example where you add other bindings

Reply:
We need it to be a vector for when we add name resolution support for `AltPattern`. The only place we should push into the vector is inside `AltPattern` see [here](https://github.com/rust-lang/rust/blob/master/compiler/rustc_resolve/src/late.rs#L3222). I'm planning to work on `AltPattern` name resolution right after this PR gets merged so I think it's better to make it a vector from now.